### PR TITLE
Add schedule emails

### LINKED ISSUE #12555: Change UI for scheduled email
### Mautic Version

5.0.x series

### PHP version

7.4, 8.0

### What browsers are you seeing the problem on?

Not relevant

### What happened?

The current way of scheduling email is confusing for many users.
The word published is used 2x:
image

Suggestion from @npracht is to have multiple steps in the UI to achieve the same result.
A new schedule button could be introduced (users are looking for it anyway)
image
Clicking on it would open a modal:
image
If you choose yes, then a "unpublish date" could be added to preserve current functionality.

### How can we reproduce this issue?

Step 1: Open an email to send
Step 2: Be confused.

### Relevant log output

```shell
None
```

### Code of Conduct

- [X] I confirm that I have read and agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by kuzmany**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes https://github.com/mautic/mautic/issues/12555  

#### Steps to test this PR:

The PR added for segment emails new schedule options. The replacement for the publish up and publish down options.

Also added new option `Continue sending email to contacts added to the segment after sending has started?` to send emails just until contact's added to segment.

image

 `Continue sending email to contacts added to the segment after sending has started?`

Now you can send emails to the segment to the contacts who were added to the segment before starting sending.

We also add two new states of current status of segment email:

Running
-  If  `Continue sending email to contacts added to the segment after sending has started?` is enabled then email If it's active it's still in state Running
- If  `Continue sending email to contacts added to the segment after sending has started?` is disabled then this state is until  pending counts are 0   (email must be active)

State

- If  `Continue sending email to contacts added to the segment after sending has started?` is disabled then this state is until  pending counts > 0  (email must be active)

Both have different states icon in the list

image

And also labels in detail (we tried to make so simple as possible but as clear as possible)

image

image

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create segment email, see like you are not able to edit publish dates in edit mode
3. Go to detail view
4. Setup schedule new options
5. Tests If works as expecteed

### CODE REVIEW

**Review by PatrickJenkner — COMMENTED** [review]
All of the added features worked just as described. However, I am still able to edit the (un-)publish date in the Edit-menu.

Screenshot 12832

**Comment by kuzmany** [comment]
@PatrickJenkner do you use dev or prod enviroment? This need `php bin/console m:a:g`

**Comment by PatrickJenkner** [comment]
@kuzmany I'm sorry, but I don't get it. From how I understood point 2. of the testing instructions, I shouldn't be able to edit the un-/publish dates in the Edit-menu. However, I can still do that. Running the command you provided doesn't seem to have changed anything.
Am I misunderstanding something?
(Also: I just tested in on GitPod, I don't know if that's dev or prod...)

**Comment by kuzmany** [comment]
@PatrickJenkner I will check it. Probably bug. Goal was remove publish dates from edit view and move it to schedule view

File: app/bundles/EmailBundle/Translations/en_US/messages.ini
Comment (RCheesley) [inline app/bundles/EmailBundle/Translations/en_US/messages.ini]:
```suggestion
mautic.email.send.datetime.end="Stop sending"
```

Grammatically speaking, I think this should be Stop Sending as you used Start sending before. If you used Begin, then you'd use End here.

File: app/bundles/EmailBundle/Translations/en_US/messages.ini
Comment (RCheesley) [inline app/bundles/EmailBundle/Translations/en_US/messages.ini]:
```suggestion
mautic.email.send.continue.tooltip="'No' will send the e-mail to whoever is in the segment at the time the e-mail is sent. 'Yes' will keep the e-mail published until the specified end date."
```

**Review by RCheesley — CHANGES_REQUESTED** [review]
A small review as I spotted some grammatical tweaks needed - plan to test tomorrow.

**Comment by atowne-mautic** [comment]
Just to verify - if the email is currently sending/in the time between Start and Stop Sending and someone selects Cancel Schedule, that will stop emails from sending right away? Will that also update the Stop Sending field? Will the Start & Stop dates also be visible in the details where we currently see Publish Up and Publish Down? 
image

Similarly, if someone updates the schedule to extend it during the time that an email is sending, that will make it so emails don't stop at the previous stop time?

Will there be any batching for contacts added to the segment after the initial send, or will the email send happen one-by-one as new contacts are added?

This looks like a great update, thank you!

**Comment by kuzmany** [comment]
@atowne-mautic  This PR just moves the publish up/publish down date fields from the edit view to the new schedule view.

Sending an email is processed with a command. There are a lot of options (batch, limits). Based on the current command setting, then we should stop sending work or processing batches.

**Comment by RCheesley** [comment]
@kuzmany I am getting the same as @PatrickJenkner - when I **create** the email _and_ when I edit, I am still seeing the dates and they are editable. It seems that's not meant to be showing as I understand it from your description in the PR?

I have rebuilt the prod assets.

screenshot-8080-kuzmany-mautic-u26vvq1de09 ws-eu107 gitpod io-2023 12 22-14_44_59

**Comment by kuzmany** [comment]
@RCheesley @PatrickJenkner Fixed with last commitł

**Comment by jos0405** [comment]
Visuals look good, but I have an error when trying to send. Is it related?
image

**Comment by kuzmany** [comment]
@jos0405 No it's not related to PR

**Comment by jos0405** [comment]
This is working as described, good to go.

**Comment by nick-vanpraet** [comment]
Also had that $limit type of int error, is indeed unrelated. Gave it limit and batch flags to bypass. The UI options seem to work as expected. As long as the checkbox to keep sending to new segments is checked and the end date hasn't passed it sent the email to any new segment members. If unchecked, it sent it to everyone in the segment and then stopped. ignoring newly added segment members.

Looks good to go to me.

**Comment by RCheesley** [comment]
@kuzmany this is good to go with a code review, but it has a conflict to resolve with the test before we can merge.

File: app/bundles/EmailBundle/Controller/EmailController.php
Comment (nick-vanpraet) [inline app/bundles/EmailBundle/Controller/EmailController.php:1372]: If continue sending is set to false, should we not empty out the publishDown date to prevent the issue where the user set it to true temporarily, picked a publish down date, then set continue sending back to false? Will prevent weird issues if we check only the publishDown date somewhere else in code.

Comment (kuzmany) [reply app/bundles/EmailBundle/Controller/EmailController.php:1372]: I'm not completely sure, but whatever we choose, it might cause confusion. I would suggest letting the users decide how they want to use it. There are different combinations available, and we have a diverse range of customers who use it for their own specific purposes.

We use it in production long time, and nobody complains about it.

**Comment by RCheesley** [comment]
@escopecz this has conflicts, could they be reviewed before we test please?

File: app/bundles/EmailBundle/Controller/EmailController.php
Comment (escopecz) [inline app/bundles/EmailBundle/Controller/EmailController.php:1324]: Please, write a functional test for this new action

File: app/bundles/EmailBundle/Translations/en_US/messages.ini
Comment (escopecz) [inline app/bundles/EmailBundle/Translations/en_US/messages.ini:283]:
```suggestion
mautic.email.send.continue.tooltip="'No' will send the e-mail to whoever is in the segment at the time the email is sent. 'Yes' will keep the email published until the specified end date."
```
As we are using "email" in the rest of the file, let's not use "e-mail".

**Comment by escopecz** [comment]
@RCheesley I bet @kuzmany will resolve them.

**Review by andersonjeccel — CHANGES_REQUESTED** [review]
From a UX perspective, it makes sense to define Yes for `Continue sending email to contacts added to the segment after sending has started?`.

But what does it mean when we select No?

`Do not continue sending email to contacts added to the segment after sending has started `
If we are not going to continue, when will we stop after starting? Will any of the segment emails be sent?

---
I would suggest polishing the sentence: 
`Send emails to contacts added to the segment after sending has started?`

File: app/bundles/EmailBundle/Controller/EmailController.php
Comment (RCheesley) [reply app/bundles/EmailBundle/Controller/EmailController.php:1324]: @kuzmany see request here

**Comment by RCheesley** [comment]
+1 on the suggestion from @andersonjeccel - some outstanding feedback to address and conflicts here @kuzmany.

**Comment by RCheesley** [comment]
Would be nice to merge this soon if we can get the conflicts and feedback addressed :)

**Comment by kuzmany** [comment]
Close favor to https://github.com/mautic/mautic/pull/14254

**Pull request closed without merging**
